File: 30-Day LeetCoding Challenge/Python/longestCommonSubsequence.py
class Solution(object):
    def longestCommonSubsequence(self, text1, text2):
        """
        :type text1: str
        :type text2: str
        :rtype: int
        """
        l1 = len(text1)
        l2 = len(text2)
        if l1 == 0 or l2 == 0:
            return 0
        ans = 0
        dp = [[0 for a in range(l2)] for b in range(l1)]
        for i in range(l1):
            for j in range(l2):
                if text1[i] == text2[j]: 
                    dp[i][j] = 1
                    if i > 0 and j > 0:
                        dp[i][j] = dp[i-1][j-1] + 1
                else:
                    if (i > 0 and j == 0):
                        dp[i][j] = dp[i-1][j]
                    elif (i == 0 and j > 0):
                        dp[i][j] = dp[i][j-1]
                    elif i > 0 and j > 0:
                        dp[i][j] = max(dp[i-1][j], dp[i][j-1])
                ans = max(ans, dp[i][j])
        return ans

File: 30-Day LeetCoding Challenge/C++/longestCommonSubsequence.cpp
class Solution {
public:
    int longestCommonSubsequence(string text1, string text2) {
        int l1 = text1.length();
        int l2 = text2.length();
        int ans = 0;
        vector<vector<int>> dp(l1, vector<int>(l2));
        for(int i = 0; i < l1; i++) {
            for(int j = 0; j < l2; j++) {
                if(text1[i] == text2[j]) {
                    dp[i][j] = 1;
                    if(i > 0 && j > 0) {
                        dp[i][j] = dp[i-1][j-1] + 1;
                    
                    }
                }
                else {
                    if(i == 0 && j > 0) {
                        dp[i][j] = dp[i][j-1];
                    }
                    else if(i > 0 && j == 0) {
                        dp[i][j] = dp[i-1][j];
                    }
                    else if(i > 0 && j > 0) {
                        dp[i][j] = max(dp[i-1][j], dp[i][j-1]);
                    }
                }
                ans = max(dp[i][j], ans);
            }
        }
        return ans;
    }
};